snuba/datasets/table_storage.py

```diff
@@ -225,6 +232,7 @@ def get_batch_writer(
         options: ClickhouseWriterOptions = None,
         table_name: Optional[str] = None,
         chunk_size: int = settings.CLICKHOUSE_HTTP_CHUNK_SIZE,
+        slice_id: Optional[int] = None,
```

Comment:
If the table writer already contains the slice_id as part of its state `self.__slice_id` why would you need to pass the slice id when instantiating the batch writer ?

```
get_cluster(self.__storage_set, self.get_slice_id()).get_batch_writer(
```

On the other hand if the same table writer is supposed to be able to return a batch writer for multiple slices I don;t think you should have a `get_slice_id` method nor `add_slice_id`.

The big problem you have with this approach (`add_slice_id` method) is that you are making talbe writers (thus storages) mutable. This can have quite substantial negative ramifications. If the storage can mutate after initialization, it means we will have to be careful, any time we reference it ,to manage properly both scenarios: before mutations and after. 
We also need to be very careful to the storage lifecycle in order not to use the storage before the slice has been added.

Immutable data models simplify things a lot specifically when these objects are conceptually singletons. Making a singleton mutable tends to introduce a lot of hard to manage side effects.

Reply:
I completely agree with Filippo's comments, especially because in this scenario we do not want `slice_id` to change within the lifetime of a process.

We should either figure out how to wire in the slice_id from the constructor (even if that means a bigger change) or require slice_id to be passed for every call of both `get_cluster` and `get_topic` (your original approach).

If you want to go with the constructor approach and you think it would help minimize complexity, I think it would be reasonable to split the work into:
- [refactor] PR 1: a mostly no-op change to take slice_id in the `TableWriter` constructor and store it, and accept that argument via the CLI
- [feature] PR 2: a functional change to actually use the new slice_id

Reply:
In order to pass in slice_id into the `TableWriter` constructor, we will also have to change the `WritableTableStorage` constructor to take in slice_id. This will mean changing the way storages will be initialized, and the way they will be built from YAML config, etc. We will then need a new alternative to the current `get_writable_storage()` in the storage factory, perhaps something like `get_sliced_writable_storage()` and changes to `_StorageFactory` as well to include sliced storages. I just wanted to make sure that these changes are expected and seem reasonable/okay for me to implement? Or perhaps is there an alternative, simpler approach that I'm missing?

Reply:
The alternative is that the storage is initialized in an agnostic way to the slice id and then the slice id is passed when you retrieve any resource that is slice specific, here there would be many possible ways depending which level should be slice specific: `get_table_writer` or within the `TableWriter` `get_batch_writer`, etc.

If instead you want to make the storage initialization slice aware, that can make all further usages simpler as no slice id parameter is needed anymore.

Which one is better depends on where and when you have the `slice_id` available, specifically you will need to look into both the query pipeline and the consumer since both rely on the storage to know where to read from and where to write to.